File: 20231014_Saturday/20231014.go
// Leetcode: 2742. Painting the Walls.
// https://leetcode.com/problems/painting-the-walls/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 54 ms, faster than 55.56% of Go online submissions for Painting the
// Walls.
// Memory Usage: 8.5 MB, less than 33.33% of Go online submissions for Painting
// the Walls.
// 2023.10.14 Daily Challenge.
package main

func paintWalls(cost []int, time []int) int {
	c := cost
	n := len(c)
	dp := make([][]int, n+1)
	inf := make([][]bool, n+1)
	for i := 0; i <= n; i += 1 {
		dp[i] = make([]int, n+1)
		inf[i] = make([]bool, n+1)
	}
	for i := 1; i <= n; i += 1 {
		inf[n][i] = true
	}
	for i := n - 1; i >= 0; i -= 1 {
		for r := 1; r <= n; r += 1 {
			tmp := max(0, r-1-time[i])

			pinf := inf[i+1][tmp]
			dinf := inf[i+1][r]

			paint := c[i] + dp[i+1][tmp]
			dont_paint := dp[i+1][r]

			if pinf && dinf {
				inf[i][r] = true
			} else if pinf {
				dp[i][r] = dont_paint
				inf[i][r] = false
			} else if dinf {
				dp[i][r] = paint
				inf[i][r] = false
			} else {
				dp[i][r] = min(paint, dont_paint)
				inf[i][r] = false
			}
		}
	}
	return dp[0][n]
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: 20231014_Saturday/20231014.rb
# Leetcode: 2742. Painting the Walls. 
# https://leetcode.com/problems/painting-the-walls/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 739 ms, faster than 100.00% of Ruby online submissions for Painting
# the Walls.
# Memory Usage: 220 MB, less than 100.00% of Ruby online submissions for
# Painting the Walls.
# 2023.10.14 Daily Challenge.
# @param {Integer[]} cost
# @param {Integer[]} time
# @return {Integer}
def paint_walls(cost, time)
  # Based on:
  # https://leetcode.com/problems/painting-the-walls/solution/
  n = cost.size
  dp = Array.new(n+1){ Array.new(n+1,0)}
  (1..n).each do |i| 
    dp[n][i] = Float::INFINITY
  end
  (0...n).reverse_each do |i|
    (1..n).each do |r|
      tmp = [0, r - 1 - time[i]].max
      paint = cost[i] + dp[i+1][tmp]
      dont_paint = dp[i+1][r]
      dp[i][r] = [paint,dont_paint].min
    end
  end
  return dp[0][n]
end
